### ISSUE TITLE
AnaphoraTest fail on clean ubuntu 16.04 (originally #2793)

### ISSUE THREAD

**Issue by mjsduncan** [opencog/opencog#2985]
```
8:AnaphoraTest

;;; compiled /home/mjsd/oc/guile/ccache/2.2-LE-8-3.A/home/mjsd/oc/opencog/tests/nlp/anaphora/data/pleonastic_it/data_#4.scm.go
;;; compiling FAIL

======================================================================
FAIL: test_conjunctions (test_anaphora.AnaphoraUnitTester)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mjsd/oc/opencog/tests/nlp/anaphora/test_anaphora.py", line 430, in test_conjunctions
    self.assertFalse(self.hobbsAgent.pleonastic_it(self.getWord('waitresses')))
AssertionError: True is not false

======================================================================
FAIL: test_pleonastic_if (test_anaphora.AnaphoraUnitTester)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mjsd/oc/opencog/tests/nlp/anaphora/test_anaphora.py", line 417, in test_pleonastic_if
    self.assertFalse(self.hobbsAgent.pleonastic_it(self.getWord('it')))
AssertionError: True is not false

======================================================================
FAIL: Testing the propose function
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mjsd/oc/opencog/tests/nlp/anaphora/test_anaphora.py", line 378, in test_propose
    filter_1()
  File "/home/mjsd/oc/opencog/tests/nlp/anaphora/test_anaphora.py", line 94, in filter_1
    self.assertTrue(self.hobbsAgent.propose(self.getWord('antecedent')))
AssertionError: False is not true

----------------------------------------------------------------------
Ran 5 tests in 9.889s

FAILED (failures=3)
accepted 
[(WordInstanceNode "b") ; [7176733813256336297][1]
, (WordInstanceNode "b") ; [7176733813256336297][1]
, (WordInstanceNode "c") ; [1954536943936637625][1]
]
accepted 
[(WordInstanceNode "c") ; [1954536943936637625][1]
, (WordInstanceNode "b") ; [7176733813256336297][1]
, (WordInstanceNode "c") ; [1954536943936637625][1]
, (WordInstanceNode "d") ; [7920996617400663969][1]
, (WordInstanceNode "e") ; [811631232188714808][1]
, (WordInstanceNode "f") ; [825601903258358495][1]
]
accepted 
[(WordInstanceNode "d") ; [7920996617400663969][1]
, (WordInstanceNode "b") ; [7176733813256336297][1]
, (WordInstanceNode "c") ; [1954536943936637625][1]
, (WordInstanceNode "d") ; [7920996617400663969][1]
, (WordInstanceNode "e") ; [811631232188714808][1]
, (WordInstanceNode "f") ; [825601903258358495][1]
, (WordInstanceNode "g") ; [7550112285569995374][1]
]
accepted 
[(WordInstanceNode "e") ; [811631232188714808][1]
, (WordInstanceNode "b") ; [7176733813256336297][1]
, (WordInstanceNode "c") ; [1954536943936637625][1]
, (WordInstanceNode "d") ; [7920996617400663969][1]
, (WordInstanceNode "e") ; [811631232188714808][1]
, (WordInstanceNode "f") ; [825601903258358495][1]
, (WordInstanceNode "g") ; [7550112285569995374][1]
]
accepted 
[(WordInstanceNode "f") ; [825601903258358495][1]
, (WordInstanceNode "b") ; [7176733813256336297][1]
, (WordInstanceNode "c") ; [1954536943936637625][1]
, (WordInstanceNode "d") ; [7920996617400663969][1]
, (WordInstanceNode "e") ; [811631232188714808][1]
, (WordInstanceNode "f") ; [825601903258358495][1]
, (WordInstanceNode "g") ; [7550112285569995374][1]
]
accepted 
[(WordInstanceNode "g") ; [7550112285569995374][1]
, (WordInstanceNode "b") ; [7176733813256336297][1]
, (WordInstanceNode "c") ; [1954536943936637625][1]
, (WordInstanceNode "d") ; [7920996617400663969][1]
, (WordInstanceNode "e") ; [811631232188714808][1]
, (WordInstanceNode "f") ; [825601903258358495][1]
, (WordInstanceNode "g") ; [7550112285569995374][1]
]
Testing filter #1...
/home/mjsd/oc/opencog/build/tests/nlp/../../../tests/nlp/anaphora/data/propose/filter-#1/data_#1.scm
;;; compiled /home/mjsd/oc/guile/ccache/2.2-LE-8-3.A/home/mjsd/oc/opencog/tests/nlp/anaphora/data/propose/filter-#1/data_#1.scm.go
<end of output>
Test time =  10.06 sec
----------------------------------------------------------
Test Failed.
"AnaphoraTest" end time: Jan 24 11:37 EST
"AnaphoraTest" time elapsed: 00:00:10
----------------------------------------------------------
```

**Comment by linas** [comment]
I looked at this a few weeks ago, and concluded that its probably failing because the python bindings to opencog are lame or busted or inoperable.  The anaphora code itself has been working a long time, and no one has messed with it, so there's no reason it should be failing .. this is a side-effect of some other subsystem.

**Comment by mjsduncan** [comment]
like maybe the python unit testing system itself is failing?

**Comment by vsbogd** [comment]
Test shows that `hobbs.py` code is broken, but after analysis it seems that it was broken from very beginning, and fixing it requires some code rework, so I would vote for skipping tests for now.

Analysis:
`BindLinkExecution.returnResults` method: https://github.com/opencog/opencog/blob/88100458896eddab2d2d39b4dfd0771ff353bc89/opencog/nlp/anaphora/agents/hobbs.py#L86-L104 always returns non-empty collection for calls like: https://github.com/opencog/opencog/blob/88100458896eddab2d2d39b4dfd0771ff353bc89/opencog/nlp/anaphora/agents/hobbs.py#L492 because it checks incoming set of the `AnchorNode("CurrentResult")`: https://github.com/opencog/opencog/blob/88100458896eddab2d2d39b4dfd0771ff353bc89/opencog/nlp/anaphora/agents/hobbs.py#L419 And this set always contains `AnchorNode("Matched")` from the `BindLink` executed: https://github.com/opencog/opencog/blob/88100458896eddab2d2d39b4dfd0771ff353bc89/opencog/nlp/anaphora/rules/isIt.scm#L21-L24 Consequently method `HobbsAgent.pleonastic_it` always returns `True` and method `HobbsAgent.propose` always returns `False`.

**Comment by linas** [comment]
But this test used to consistently pass ... why does it no longer pass?

**Comment by linas** [comment]
Yeah, the test is insane; `opencog/nlp/anaphora/rules/isIt.scm` creates an obviously-faulty bind-link, which always evaluates to true.  I'm utterly stumped as to why this test ever passed... there must have been some counter-vailing bug somewhere else!?

**Comment by linas** [comment]
I disabled the unit test in #3525 Apparently, this code never worked. Apparently, ghost does not use it, either.  Go figure.  Its abandon-ware.

**Comment by linas** [comment]
Right #3524 is better than #3525. Applied; thanks.

**Comment by linas** [comment]
Closing, fixed in #3527 -- The tests no longer have to be skipped, the full unit test runs to completion and passes. 

The original developer of teh code got mostly everything right, in that it seems to have the right idea, and most of the code was right, but then .. bizarrely misunderstood bindlinks.  I fixed all that; it seems to be working.
